src/batchmap.rs

```diff
@@ -0,0 +1,627 @@
+        self.keys = Some(keys);
+        self
+    }
+
+    /// Sets or replaces the tags associated with this message.
+    ///
+    /// # Arguments
+    ///
+    /// * `tags` - A vector of strings representing the tags.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use numaflow::batchmap::Message;
+    /// let message = Message::new(vec![1, 2, 3]).tags(vec!["tag1".to_string(), "tag2".to_string()]);
+    /// ```
+    pub fn tags(mut self, tags: Vec<String>) -> Self {
+        self.tags = Some(tags);
+        self
+    }
+
+    /// Replaces the value of the message.
+    ///
+    /// # Arguments
+    ///
+    /// * `value` - A new vector of bytes that replaces the current message value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use numaflow::batchmap::Message;
+    /// let message = Message::new(vec![1, 2, 3]).value(vec![4, 5, 6]);
+    /// ```
+    pub fn value(mut self, value: Vec<u8>) -> Self {
+        self.value = value;
+        self
+    }
+}
+/// The result of the call to [`BatchMapper::batchmap`] method.
+pub struct BatchResponse {
+    /// id is the unique ID of the message.
+    pub id: String,
+    // message is the response from the batch map.
+    pub message: Vec<Message>,
+}
+
+impl BatchResponse {
+    /// Creates a new `BatchResponse` for a given id and empty message.
+    pub fn from_id(id: String) -> Self {
+        Self {
+            id,
+            message: Vec::new(),
+        }
+    }
+
+    /// append a message to the response.
+    pub fn append(&mut self, message: Message) {
+        self.message.push(message);
+    }
+}
+
+impl From<Message> for proto::batch_map_response::Result {
+    fn from(value: Message) -> Self {
+        proto::batch_map_response::Result {
+            keys: value.keys.unwrap_or_default(),
+            value: value.value,
+            tags: value.tags.unwrap_or_default(),
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl<T> BatchMap for BatchMapService<T>
+where
+    T: BatchMapper + Send + Sync + 'static,
+{
+    async fn is_ready(
+        &self,
+        _: Request<()>,
+    ) -> Result<tonic::Response<proto::ReadyResponse>, Status> {
+        Ok(tonic::Response::new(proto::ReadyResponse { ready: true }))
+    }
+
+    type BatchMapFnStream = ReceiverStream<Result<proto::BatchMapResponse, Status>>;
+
+    async fn batch_map_fn(
+        &self,
+        request: Request<Streaming<proto::BatchMapRequest>>,
+    ) -> Result<Response<Self::BatchMapFnStream>, Status> {
+        let mut stream = request.into_inner();
+        let (tx, rx) = mpsc::channel::<Datum>(1);
+
+        // Create a channel to send the response back to the grpc client.
+        let (grpc_response_tx, grpc_response_rx) =
+            channel::<Result<proto::BatchMapResponse, Status>>(1);
+
+        let shutdown_tx = self._shutdown_tx.clone();
+
+        // counter to keep track of the number of messages received
+        let counter_orig = Arc::new(AtomicUsize::new(0));
```

Comment:
we dont need this, there is a natural barrier.. make your spawn return the count and compare that in the next

Reply:
In the case where the user returns from the UDF function early before consuming all the messages on the channel, in such a scenario we would need this right?

Reply:
That should be fine, because you are awaiting on the spawned task so it won't exit.. perhaps do a `select!` to force the selection since returning early is not right.